2022 Agent Key Rotation docs (#11803)
DataDog/documentation: content/en/agent/faq/linux-agent-2022-key-rotation.md

kind: faq
---

As a common best practice, Datadog periodically rotates the keys and certificates used to sign Datadog's Agent packages. The following GPG keys, used to sign the Agent RPM and DEB packages, reach their end-of-life in June 2022 and will be rotated in April 2022:

- The RPM signing key with hash [`A4C0B90D7443CF6E4E8AA341F1068E14E09422B3`][1] will be rotated in April 4th 2022 and will be replaced by the key with hash [`C6559B690CA882F023BDF3F63F4D1729FD4BF915`][2]
- The DEB signing key with hash [`A2923DFF56EDA6E76E55E492D3A80E30382E94DE`][3] will be rotated in April 18th 2022 and will be replaced by the key with hash [`D75CEA17048B9ACBF186794B32637D44F14F620E`][4]

Customers using Datadog's RPM or DEB packages might require a manual action to import the new key on their systems in order to install or upgrade the Agent after the rotation takes place.

<div class="alert alert-info">
<strong>Note</strong>: This DOES NOT affect the functionality of already running Agents, and only limits the ability to install or upgrade to a newer version of the Agent. This also doesn't affect Windows or MacOS Agents, nor Dockerized Linux Agents.
</div>

> If you're using one of the following install methods, your hosts trust the key automatically and no further action is needed:

Reviewer: Maybe add a title here such as - "## Install methods that automatically trust the new PGP key"

* [Agent install script][5] v1.6.0 or later (released Jul 26, 2021)
* [Chef cookbook][6] v4.11.0 or later (released Aug 10, 2021)
* [Ansible role][7] v4.10.0 or later (released May 25, 2021)
* [Puppet module][8] v3.13.0 or later (released Aug 11, 2021)
* [SaltStack formula][9] v3.4 or later (released Aug 12, 2021)
* [Heroku buildpack][10] v1.26  or later (released May 26, 2021)
* [Elastic Beanstalk][11] config templates updated as of Mar 29, 2021 or later (should contain `DATADOG_RPM_KEY_FD4BF915.public` under `gpgkey`)
* Containerized Agents (Docker/Kubernetes): No action needed on any version
* Windows/MacOS Agents: No action needed on any version.

Additionally, if you are using the DEB version of the Agent version 7.31.0 or greater, your hosts should have the `datadog-signing-keys` package installed, which will make APT recognize the signature by the new key on the Datadog APT repository.

> If unsure, read the steps in [Check if a host trusts the new GPG key](#check-if-a-host-trusts-the-new-gpg-key).

Reviewer: If you're unsure if a host trusts the new PGP key, you can [check](#check-if-a-host-trusts-the-new-gpg-key).

Trying to install or upgrade the Agent package without trusting the new key results in `NOKEY` errors when installing the RPM package and `NO_PUBKEY` errors when installing the DEB package.

> For hosts running older versions of the install methods listed above or older versions of the DEB package, Datadog recommends updating the install method to the latest version. Alternatively Debian and Ubuntu users can update the Agent to version 7.31.0 or greater. Otherwise, the keys can be manually updated on host by following these instructions:

Reviewer: "... can be manually updated."

> ### Manual update

Reviewer: I think this should be a ## Manual Update 
That way the heading of this and the next paragraph are the same

---
title: 2022 Linux Agent Key Rotation
kind: faq
---

As a common best practice, Datadog periodically rotates the keys and certificates used to sign Datadog's Agent packages. The following GPG keys, used to sign the Agent RPM and DEB packages, reach their end-of-life in June 2022 and will be rotated in April 2022:

- The RPM signing key with hash [`A4C0B90D7443CF6E4E8AA341F1068E14E09422B3`][1] will be rotated in April 4th 2022 and will be replaced by the key with hash [`C6559B690CA882F023BDF3F63F4D1729FD4BF915`][2]

> - The DEB signing key with hash [`A2923DFF56EDA6E76E55E492D3A80E30382E94DE`][3] will be rotated in April 18th 2022 and will be replaced by the key with hash `D75CEA17048B9ACBF186794B32637D44F14F620E`[4]

Reviewer 2: ```suggestion
- The DEB signing key with hash [`A2923DFF56EDA6E76E55E492D3A80E30382E94DE`][3] will be rotated in April 18th 2022 and will be replaced by the key with hash [`D75CEA17048B9ACBF186794B32637D44F14F620E`][4]
```

Customers using Datadog's RPM or DEB packages might require a manual action to import the new key on their systems in order to install or upgrade the Agent after the rotation takes place.

<div class="alert alert-info">
**Note**: This DOES NOT affect the functionality of already running Agents, and only limits the ability to install or upgrade to a newer version of the Agent. This also doesn't affect Windows or MacOS Agents, nor Dockerized Linux Agents.
</div>

If you're using one of the following install methods, your hosts trust the key automatically and no further action is needed:

* [Agent install script][5] v1.5.0 (released Jun 29, 2021)
* [Chef cookbook][6] v4.11.0 (released Aug 10, 2021)
* [Ansible role][7] v4.9.0 (released May 6, 2021)
* [Puppet module][8] v3.13 (released Aug 11, 2021)

> * [SaltStack formula][9] v3.4 (released Aug 12, 2021)

Reviewer 2: We should say "vX.Y.Z or later" here

Reviewer 2: >  Maybe we could modify the sentence to make it clearer (e.g. "starting from version X")

That also works for me. In general I think one needs to be as explicit and precise as possible in docs

Author: ```suggestion
* [Agent install script][5] v1.5.0 or later (released Jun 29, 2021)
* [Chef cookbook][6] v4.11.0 or later (released Aug 10, 2021)
* [Ansible role][7] v4.9.0 or later (released May 6, 2021)
* [Puppet module][8] v3.13 or later (released Aug 11, 2021)
* [SaltStack formula][9] v3.4 or later (released Aug 12, 2021)
```

* [Elastic Beanstalk][10] config templates updated as of Mar 29, 2021 (should contain `DATADOG_RPM_KEY_FD4BF915.public` under `gpgkey`)
* Containerized Agents (Docker/Kubernetes): No action needed on any version
* Heroku: No action needed on any version.
* Windows/MacOS Agents: No action needed on any version.

> Additionally, if you are using the DEB version of the Agent version 7.31.0 or greater, your hosts should have installed the `datadog-signing-keys` package that automatically trusts the new key.

Reviewer 2: Why not add this as a bullet point to the list?

Author: Because it overlaps: you can be using an older chef but a newer Agent, or the other way around, and still be fine. How would you include it in the list above in a way that the overlap is clear?

Reviewer 2: That's a fair point, we can leave it as a separate sentence. I didn't get/think about the overlap when reading it like this, but probably that is just me

If unsure, read the steps in [Check if a host trusts the new GPG key](#check-if-a-host-trusts-the-new-gpg-key).

Trying to install or upgrade the Agent package without trusting the new key results in `NOKEY` errors when installing the RPM package and `NO_PUBKEY` errors when installing the DEB package.

For hosts running older versions of the install methods listed above or older versions of the DEB package, Datadog recommends updating the install method to the latest version. Alternatively, Debian and Ubuntu users can update the Agent to version 7.31.0 or greater. Otherwise, the keys can be manually updated on host by following these instructions:

### Manual update

{{< tabs >}}
{{% tab "Debian/Ubuntu" %}}

Run the following commands on the host:

```bash

> $ wget -qO - https://keys.datadoghq.com/DATADOG_APT_KEY_F14F620E.public | sudo apt-key add -

Reviewer 2: Maybe split into two steps and check the key hash before adding?

Reviewer 2: No strong feelings about this so I am approving

> $ wget -qO - https://keys.datadoghq.com/DATADOG_APT_KEY_F14F620E.public | sudo apt-key add -

Reviewer 3: So while this would work, technically it's not ideal because the usage of `apt-key` for 3rd part repos is discouraged per https://wiki.debian.org/DebianRepository/UseThirdParty#OpenPGP_Key_distribution

I understand that we're doing this to accomodate for people who haven't yet adopted the approach of using `signed-by` option for the Datadog DEB repository, so I guess it's ok to keep this, but we should at least provide a clear note about this not being a good way to go. Alternatively, we could provide link to the in-app installation instructions at https://app.datadoghq.com/account/settings#agent/debian which do the whole thing with keyring and `signed-by`.

Author: The problem with the better way is that the `datadog.list` contents change depending on the Agent major version, and that would make the docs quite more complex... futhermore, if you have use a config management tool, it will rewrite the file on the next run.

And now that I think about it, this is a problem for RPM as well :/

Reviewer 3: Hmm, that's a good point. I think this should be highlighted in the docs - if you're using an older version of a cfg management tool, your hand changes will be overwritten.

In terms of RPM platforms, importing the key to the `rpm` database will work for package-level signatures. For repodata signatures, it will work on SUSE, because zypper respects rpm-imported GPG keys when validating repodata signatures. However, for `yum`/`dnf` this is not true - people who modify repofile manually on a RH platform, but then rerun a cfg management tool and clear yum/dnf cache will not have the new key for repodata verification.

Author: On the other hand, repodata validation is only turned on by install methods that already add the `CURRENT` key to the list of keys, so we don't need to worry about that. I think we can just `rpm --import` (but we would need to test all distros and different versions of config management tools...).

Reviewer 3: Oh, that's right! Well, at least that makes it easier for us :)

```

{{% /tab %}}
{{% tab "RedHat/CentOS" %}}

Edit the `/etc/yum.repos.d/datadog.repo` and make sure it includes the following list of keys in the `gpgkey` section:

```
gpgkey=https://keys.datadoghq.com/DATADOG_RPM_KEY_CURRENT.public
       https://keys.datadoghq.com/DATADOG_RPM_KEY_FD4BF915.public
       https://keys.datadoghq.com/DATADOG_RPM_KEY_E09422B3.public
       https://keys.datadoghq.com/DATADOG_RPM_KEY.public
```

{{% /tab %}}
{{% tab "SUSE" %}}

> Edit the `/etc/zypp/repos.d/datadog.repo` and make sure it includes the following list of keys in the `gpgkey` section:

Reviewer 3: I think we should provide a comment about the fact that for SUSE 11 this entry can only contain one key (the `CURRENT` one).

```
gpgkey=https://keys.datadoghq.com/DATADOG_RPM_KEY_CURRENT.public
       https://keys.datadoghq.com/DATADOG_RPM_KEY_FD4BF915.public
       https://keys.datadoghq.com/DATADOG_RPM_KEY_E09422B3.public
       https://keys.datadoghq.com/DATADOG_RPM_KEY.public
```

{{% /tab %}}
{{< /tabs >}}

## Check if a host trusts the new GPG key

{{< tabs >}}
{{% tab "Debian/Ubuntu" %}}

Run the following commands on the host:

```bash

> $ apt-key list

Reviewer 3: If we keep doing `apt-key add` above, then this is true; otherwise if we advise to use the separate keyring, this will need to be `gpg --no-default-keyring --keyring /usr/share/keyrings/datadog-archive-keyring.gpg --list-keys`.

Author: This should actually ask you to run both commands, because you could have the key added either way.

Reviewer 3: Thinking about this further, I *think* that all cfg management tools versions that create the separate keyring also import the new key into it, but you're right that if people want to verify, they need to be provided both versions of what to do based on how they set the system up.

Author: I've left it just as `rpm -q` on rpm-based and `apt-key list` on debian-based. Once I test it I'll check if it works on all distros (it will only work if we always add the key to the "generic" keyring instead of adding it per-repo).

---
title: 2022 Linux Agent Key Rotation
kind: faq
---

As a common best practice, Datadog periodically rotates the keys and certificates used to sign Datadog's Agent packages. The following GPG keys, used to sign the Agent RPM and DEB packages, reach their end-of-life in June 2022 and will be rotated in April 2022:

- The RPM signing key with hash [`A4C0B90D7443CF6E4E8AA341F1068E14E09422B3`][1] will be rotated in April 11th 2022 and will be replaced by the key with hash [`C6559B690CA882F023BDF3F63F4D1729FD4BF915`][2]

> - The DEB signing key with hash [`A2923DFF56EDA6E76E55E492D3A80E30382E94DE`][3] will be rotated in May 2nd 2022 and will be replaced by the key with hash [`D75CEA17048B9ACBF186794B32637D44F14F620E`][4]

Reviewer 4: ```suggestion
- The DEB signing key with hash [`A2923DFF56EDA6E76E55E492D3A80E30382E94DE`][3] will be rotated on May 2, 2022 and replaced by the key with hash [`D75CEA17048B9ACBF186794B32637D44F14F620E`][4]
```

> Customers using Datadog's RPM or DEB packages might require a manual action to import the new key on their systems in order to install or upgrade the Agent after the rotation takes place.

Reviewer 4: ```suggestion
Customers using Datadog's RPM or DEB packages might require a manual action to import the new key on their systems to install or upgrade the Agent after the rotation takes place.
```

<div class="alert alert-info">

> <strong>Note</strong>: This DOES NOT affect the functionality of already running Agents, and only limits the ability to install the Agent or upgrade to a newer version of the Agent. This also doesn't affect Dockerized Linux Agents nor Windows or MacOS Agents.

Reviewer 4: ```suggestion
<strong>Note</strong>: This DOES NOT affect the functionality of already running Agents, and only limits the ability to install or upgrade to a newer version of the Agent. Also, this doesn't affect Dockerized Linux Agents, Windows, or macOS Agents.
```

</div>

## Install methods that automatically trust the new GPG key

> If you're using one of the following install methods, your hosts trust the key automatically and no further action is needed:

Reviewer 4: ```suggestion
Your host automatically trusts the new key (no further action is required) if you're using one of the following install methods:
```

> * [Agent install script][5] v1.6.0 or later (released Jul 26, 2021)

Reviewer 4: ```suggestion
- [Agent install script][5] v1.6.0+ (released Jul 26, 2021)
```

> * [Chef cookbook][6] v4.11.0 or later (released Aug 10, 2021)

Reviewer 4: ```suggestion
- [Chef cookbook][6] v4.11.0+ (released Aug 10, 2021)
```

> * [Ansible role][7] v4.10.0 or later (released May 25, 2021)

Reviewer 4: ```suggestion
- [Ansible role][7] v4.10.0+ (released May 25, 2021)
```

> * [Puppet module][8] v3.13.0 or later (released Aug 11, 2021)

Reviewer 4: ```suggestion
- [Puppet module][8] v3.13.0+ (released Aug 11, 2021)
```

> * [SaltStack formula][9] v3.4 or later (released Aug 12, 2021)

Reviewer 4: ```suggestion
- [SaltStack formula][9] v3.4+ (released Aug 12, 2021)
```

> * [Heroku buildpack][10] v1.26 or later (released May 26, 2021)

Reviewer 4: ```suggestion
- [Heroku buildpack][10] v1.26+ (released May 26, 2021)
```

> * [Elastic Beanstalk][11] config templates updated as of Mar 29, 2021 or later (should contain `DATADOG_RPM_KEY_FD4BF915.public` under `gpgkey`)

Reviewer 4: ```suggestion
- [Elastic Beanstalk][11] config templates updated as of Mar 29, 2021 or later (should contain `DATADOG_RPM_KEY_FD4BF915.public` under `gpgkey`)
```

> * Containerized Agents (Docker/Kubernetes): No action needed on any version

Reviewer 4: ```suggestion
- Containerized Agents (Docker/Kubernetes) for any version
```

> * Windows/MacOS Agents: No action needed on any version.

Reviewer 4: ```suggestion
- Windows/MacOS Agents for any version
```

> Additionally, if you are using the DEB version of the Agent version 7.31.0 or greater, your hosts should have the `datadog-signing-keys` package installed, which will make APT recognize the signature by the new key on the Datadog APT repository.

Reviewer 4: ```suggestion
Additionally, if you are using DEB Agent v7.31.0+, your hosts should have the `datadog-signing-keys` package installed, which makes APT recognize the signature by the new key on the Datadog APT repository.
```

If you're unsure if a host trusts the new signing key, you can [check](#check-if-a-host-trusts-the-new-gpg-key).

> For hosts running older versions of the install methods listed above or older versions of the DEB package, Datadog recommends updating the install method to the latest version. Alternatively Debian and Ubuntu users can update the Agent to version 7.31.0 or greater. Otherwise, the key can be [manually updated](#manual-update).

Reviewer 4: ```suggestion
For hosts running older versions of the install methods listed above or older versions of the DEB package, Datadog recommends updating the install method to the latest version. Alternatively Debian and Ubuntu users can update the Agent to version 7.31.0+. Otherwise, the key can be [manually updated](#manual-update).
```

## What happens if I don't trust the new key before it is rotated?

{{< tabs >}}
{{% tab "Debian/Ubuntu" %}}

Trying to install or upgrade Agent packages without trusting the new key results in `NO_PUBKEY` errors. This applies to both newly released and existing versions of the Agent.

```
The following signatures couldn't be verified because the public key is not available: NO_PUBKEY
```

{{% /tab %}}
{{% tab "RedHat/CentOS/SUSE" %}}

Installing new versions of the Agent released since April 2022 causes `NOKEY` errors. Older versions of the Agent can still be installed.

```
The GPG keys listed for the "Datadog, Inc." repository are already installed but they are not correct for this package.
Check that the correct key URLs are configured for this repository.
```

{{% /tab %}}
{{< /tabs >}}

## Manual update

{{< tabs >}}
{{% tab "Debian/Ubuntu" %}}

Run the following commands on the host:

```bash
$ curl -o /tmp/DATADOG_APT_KEY_F14F620E https://keys.datadoghq.com/DATADOG_APT_KEY_F14F620E.public
$ sudo apt-key add /tmp/DATADOG_APT_KEY_F14F620E

> $ cat /tmp/DATADOG_APT_KEY_F14F620E | sudo gpg --import --batch --no-default-keyring --keyring /usr/share/keyrings/datadog-archive-keyring.gpg

Reviewer 3: This is the scenario which assumes that the keyring file doesn't exist yet. I just tested it and running the command fails when the keyring doesn't exist. We should add `touch /usr/share/keyrings/datadog-archive-keyring.gpg` before this command.

$ sudo chmod a+r /usr/share/keyrings/datadog-archive-keyring.gpg
```

{{% /tab %}}
{{% tab "RedHat/CentOS/SUSE" %}}

Run the following commands on the host:

```
$ curl -o /tmp/DATADOG_RPM_KEY_FD4BF915 https://keys.datadoghq.com/DATADOG_RPM_KEY_FD4BF915.public
$ sudo rpm --import /tmp/DATADOG_RPM_KEY_FD4BF915
```

{{% /tab %}}
{{< /tabs >}}

## Check if a host trusts the new GPG key

{{< tabs >}}
{{% tab "Debian/Ubuntu" %}}

If the file `/usr/share/keyrings/datadog-archive-keyring.gpg` exists, the new key is trusted and no further action is needed.

{{% /tab %}}
{{% tab "RedHat/CentOS/SUSE" %}}

> Run the following commands on the host:

Reviewer 4: ```suggestion
Run the following command on the host:
```

```bash
$ rpm -q gpg-pubkey-fd4bf915
```

If the key is trusted, the command has a 0 exit code and outputs:

```
gpg-pubkey-fd4bf915-5f573efe
```

Otherwise, the command returns a non-0 exit code and the following output:

```
package gpg-pubkey-fd4bf915 is not installed
```

{{% /tab %}}
{{< /tabs >}}

> ## Impact for Agent 5 users

Reviewer 4: ```suggestion
## Impact for Agent v5 users
```

> Agent 5 users on DEB-based systems (Debian/Ubuntu) are also required to trust the new signing key in order to install or upgrade the Agent after the rotation date. Agent 5 users on RPM-based systems (RedHat/CentOS/SUSE) are not affected by this rotation.

Reviewer 4: ```suggestion
Agent v5 users on DEB-based systems (Debian/Ubuntu) are also required to trust the new signing key to install or upgrade the Agent after the rotation date. Agent v5 users on RPM-based systems (RedHat/CentOS/SUSE) are not affected by this rotation.
```

> Note that Agent 5 uses Python 2 which reached end-of-life on January 1st 2021. We recommend you [upgrade to Agent 7][12].

Reviewer 4: ```suggestion
**Note**: Agent v5 uses Python 2 which reached end-of-life on January 1, 2021. Datadog recommends [upgrading to Agent v7][12].
```

## What happens if I don't trust the new key before it is rotated?

{{< tabs >}}
{{% tab "Debian/Ubuntu" %}}

Trying to install or upgrade Agent packages without trusting the new key results in `NO_PUBKEY` errors. This applies to both newly released and existing versions of the Agent.

```
The following signatures couldn't be verified because the public key is not available: NO_PUBKEY
```

{{% /tab %}}
{{% tab "RedHat/CentOS/SUSE" %}}

Installing new versions of the Agent released since April 2022 causes `NOKEY` errors. Older versions of the Agent can still be installed.

```
The GPG keys listed for the "Datadog, Inc." repository are already installed but they are not correct for this package.
Check that the correct key URLs are configured for this repository.
```

{{% /tab %}}
{{< /tabs >}}

## Manual update

{{< tabs >}}
{{% tab "Debian/Ubuntu" %}}

Run the following commands on the host:

```bash
$ curl -o /tmp/DATADOG_APT_KEY_F14F620E https://keys.datadoghq.com/DATADOG_APT_KEY_F14F620E.public

> $ sudo apt-key add /tmp/DATADOG_APT_KEY_F14F620E

Reviewer 3: We might add a comment saying that this is only necessary on Debian < 9 and Ubuntu < 16, but it's not strictly necessary.

Author: I think it's better to keep it simple and not add more conditionals.

Reviewer 3: I agree with not adding conditionals. I just think users might wonder why the key has to be added twice, so a comment explaining it could be helpful.

Author: It's a bit hard to explain because it's a bit more confusing than "only necessary on Debian < 9 and Ubuntu < 16". 

Actually, the only part that's needed on all distros is the `apt-key add` because the keyring file is not going to be used by the `sources.list.d` file: if it is used it means it was created with an install method that also created the keyring, so this is not needed.

The only reason I'm adding the keyring file is because the way we are telling people to check if a system is in a good/bad state is to check for the file. The keyring file could even be empty and everything would still hold.

Reviewer 3: Oh you're right, we would need to modify the sources.list.d file, which would be even more complicated, so it's probably best to use apt-key add 👍

```

{{% /tab %}}
{{% tab "RedHat/CentOS/SUSE" %}}

Edit the `/etc/yum.repos.d/datadog.repo` and make sure it includes the following list of keys in the `gpgkey` section:

> {{% tab "" %}}

Reviewer 3: Should this tab have empty title?

Author: Oops

$ apt-key list
```

And check the output for Datadog's key fingerprint ending in `F14F620E` or (with a space) `F14F 620E`.

{{% /tab %}}
{{% tab "RedHat/CentOS/SUSE" %}}

Run the following commands on the host:

```bash
$ rpm -q gpg-pubkey-fd4bf915
```

If the key is trusted, the command has a 0 exit code and outputs:

```

> gpg-pubkey-e09422b3-57744e9e

Reviewer 3: Isn't this the old key? For me, the operation actually outputs `gpg-pubkey-fd4bf915-5f573efe`.

Author: Indeed.

```

Otherwise, the command returns a non-0 exit code and the following output:

```

> package gpg-pubkey-e09422b3 is not installed

Reviewer 3: Should this be `gpg-pubkey-fd4bf915`?
